storage/replicate_queue.go

```diff
@@ -75,21 +75,35 @@ func (rq replicateQueue) shouldQueue(now proto.Timestamp, repl *Replica) (should
 		return
 	}
 
-	return rq.needsReplication(zone, repl, repl.Desc())
+	delta := rq.replicaDelta(zone, repl, repl.Desc())
+	if delta == 0 {
+		if log.V(1) {
+			log.Infof("%s has the correct number of nodes", repl)
+		}
+		return false, 0
+	}
+	if delta > 0 {
+		if log.V(1) {
+			log.Infof("%s needs to add %d nodes", repl, delta)
+		}
+		// For ranges which need additional replicas, increase the priority
+		return true, float64(delta + 10)
+	}
+	if log.V(1) {
+		log.Infof("%s needs to remove %d nodes", repl, 0-delta)
+	}
+	// For ranges which have too many replicas, priority is absolute value of
+	// the delta.
+	return true, float64(0 - delta)
```

Comment:
3/5 is also more fragile than 2/3.
On Sep 4, 2015 7:01 PM, "Bram Gruneir" <notifications@github.com> wrote:

> In storage/replicate_queue.go
> <https://github.com/cockroachdb/cockroach/pull/2350#discussion_r38801281>:
>
> > +		}
> > +		return false, 0
> > +	}
> > +	if delta > 0 {
> > +		if log.V(1) {
> > +			log.Infof("%s needs to add %d nodes", repl, delta)
> > +		}
> > +		// For ranges which need additional replicas, increase the priority
> > +		return true, float64(delta + 10)
> > +	}
> > +	if log.V(1) {
> > +		log.Infof("%s needs to remove %d nodes", repl, 0-delta)
> > +	}
> > +	// For ranges which have too many replicas, priority is absolute value of
> > +	// the delta.
> > +	return true, float64(0 - delta)
>
> I don't see why you don't address it right now. We have a viable solution
> already.
> The attributes don't really come into play here.
> What matters is how brittle the range is. And one with 4/5 replicas can
> still lose one more replica before becoming unavailable. One that is at 3/5
> or 2/3 is in more danger than the 4/5 one. I'd be down if you want to
> prioritize a 3/5 over a 2/3 one, as you suggest it probably is because the
> data is more important.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2350/files#r38801281>.
>